File: SiderShopping/app/src/main/java/com/mmart/sidershopping/CategoryAdapter.java
package com.mmart.sidershopping;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.squareup.picasso.Picasso;

import java.util.ArrayList;


/**
 * Created by User on 2/12/2018.
 */

public class CategoryAdapter extends RecyclerView.Adapter<CategoryAdapter.ViewHolder> {

    private static final String TAG = "RecyclerViewAdapter";

    //vars
    private ArrayList<Category> mCategories = new ArrayList<>();
    private Context mContext;

    public CategoryAdapter(ArrayList<Category> categories, Context context) {
        mCategories = categories;
        mContext = context;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.category_item_list, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, final int position) {
        final Category category = mCategories.get(position);

        Picasso.get()
                .load(category.getmImage())
                .fit()
                .centerCrop()
                .into(holder.image);

        holder.name.setText(category.getmNome());

        holder.image.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
//                Log.d(TAG, "onClick: clicked on an image: " + mNames.get(position));
                Toast.makeText(mContext, category.getmNome(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    public int getItemCount() {
        return mCategories.size();
    }

    public class ViewHolder extends RecyclerView.ViewHolder{

        ImageView image;
        TextView name;

        public ViewHolder(View itemView) {
            super(itemView);
            image = itemView.findViewById(R.id.image_view);
            name = itemView.findViewById(R.id.name);
        }
    }
}

File: app/src/main/java/br/com/amedigital/category/CategoryAdapter.kt
package br.com.amedigital.category

import android.support.constraint.ConstraintLayout
import android.support.v7.widget.RecyclerView
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import br.com.amedigital.R
import br.com.amedigital.model.Category
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.item_category.view.*

class CategoryAdapter (private var categories: List<Category>?,
                       private val listener: (Int) -> Unit) : RecyclerView.Adapter<CategoryAdapter.ViewHolder>() {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.item_category, parent, false)
        return ViewHolder(view)
    }

    override fun getItemCount(): Int {
        return categories?.size ?: 0
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val category = categories?.get(position)
        holder.bind(category, position, listener)
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {

        private val constraintLayoutContent: ConstraintLayout = itemView.constraintLayoutContent
        private val imageViewCategory : ImageView = itemView.imageviewCategory
        private val textViewDescriptionProd : TextView = itemView.textViewDescrProd

        fun bind(category: Category?, position: Int, listener: (Int) -> Unit) {
            Log.i(javaClass.simpleName, category?.urlImage)
            category?.apply {
                Picasso.get()
                    .load(urlImage)
                    .placeholder(R.mipmap.image_place_holder)
                    .into(imageViewCategory)
                textViewDescriptionProd.text = description
            }

            constraintLayoutContent.setOnClickListener {
                listener(position)
            }
        }
    }
}